Customize how (test) is run using (action)

emillon opened the pull request.

emillon committed:
Customize how (test) is run using (action)

Closes #1195

Signed-off-by: Etienne Millon <me@emillon.org>

diml commented on src/gen_rules.ml:
```diff
@@ -164,51 +164,44 @@ module Gen(P : Install_rules.Params) = struct
+        | Some a -> a
+        | None -> Action.Unexpanded.Run (String_with_vars.make_text loc test_exe, [])
+      in
+      let extra_bindings = Pform.Map.singleton "test" (Pform.Var.Values [Value.String test_exe]) in
```
The value should be a `Path` so that it expands to the right path if the user does: `(chdir %{project_root} (run %{test}))`. Additionally, it should resolve to the executable in the host build context in case of cross-compilation (the function to do that is `Super_context.Action.map_exe`).

emillon commented:
I fixed this. I changed it so that the default action is `(run %{test})`, which takes case of this `chdir` and cross compilation issues for the default case as well.

rgrinberg commented on src/gen_rules.ml:
```diff
@@ -164,51 +164,46 @@ module Gen(P : Install_rules.Params) = struct
-        } in
+        match t.action with
+        | Some a -> a
+        | None -> Action.Unexpanded.Run (String_with_vars.virt_var __POS__ test_var_name, [])
```
Why use a fake loc here? Isn't it better to inherit the loc of the test name?

emillon commented:
good idea! just added `make_var` to do that.

rgrinberg reviewed (approved):
Nice change!
